### PR TITLE
possible fix for prevent segfault on v8

### PR DESCRIPTION
**Pull request by devzer01**
@mscdex Brian, i am not an expert on the topic. but with my limited knowledge and few hours of GDBing through the race condition, i am suggesting this fix to bindings.cc. this should handle any case where Client is garbage collected before the call ends on bindings and resulting a callback to a none existing JS object. 

all relevant info was added to PR #130

### CODE REVIEW

**Comment by devzer01** [comment]
i realized my change to bindings caused onclose to never fire. so i removed it and set the null handler at _onclose;

**Comment by devzer01** [comment]
@mscdex looks like that fixed it. :)

**Comment by mscdex** [comment]
It looks like there's still a race condition of some kind: https://travis-ci.org/mscdex/node-mariasql/jobs/110561898

**Comment by devzer01** [comment]
No problem. Let me try to reproduce it. Thanks

**Comment by devzer01** [comment]
can't reproduce on 5.x even with forcing gc, going to build 0.10.x from source and try. will let you know.

**Comment by devzer01** [comment]
not reproduce-able on 10.x (latest) on Ubuntu. is it only with OSX?

**Comment by mscdex** [comment]
I'm not sure, but that was the same configuration that had the problem the last time. I had to re-run it a couple of times before it happened though.

**Comment by devzer01** [comment]
I will try to reproduce it on my Mac. I looked at the failing test case and I am yet to figure out why it would fail there. Because now I set handle to null at onclose. If i was setting null on the wrong_handle it would have been reproducible regardless of the os version. I will try to modify the test a bit and see if I can reproduce it by calling gc() from the test

**Comment by devzer01** [comment]
do you know what is the exact version of 0.10.x? you are running on travis? because

```
# Fatal error in ../deps/v8/src/api.h, line 297
# CHECK(allow_empty_handle || that != __null) failed
```

in latest 0.10.x line 297 is not that 

its 

```
OPEN_HANDLE_LIST(MAKE_OPEN_HANDLE)
```

**Comment by devzer01** [comment]
@mscdex Brian what do you think? looks green again i will add some comments to the branch to refresh the build process and see if we have race-condition still.

**Comment by devzer01** [comment]
@mscdex looks like we got two green in a row.

**Comment by devzer01** [comment]
@mscdex Brian can we merge this now please :)

**Comment by devzer01** [comment]
@mscdex ping :) ?

**Comment by mscdex** [comment]
Sorry, I've been a bit busy. I hope to get time to look at this again soon.

**Comment by devzer01** [comment]
no worries, i think the latest changes i made now is stable. setting `_handle = null` on `process.nextTicket()` at the end of `onclose` seems to solve the problem of C-bindings segfaulting. also was there any reason you didn't directly implement a `kill` method in c-bindings rather than using the Class it self to create another connection to kill an existing connection.

**Comment by mscdex** [comment]
With regard to killing, `mysql_kill()` is actually deprecated and the documentation actually suggests uses the `KILL()` query instead. Also `mysql_kill()` does not handle >32bit connection/thread ids, which is probably why it's deprecated.

**Comment by devzer01** [comment]
i mean having 
```
killConnection(int ID) {
mysql_init(); mysql_connect(); mysql_query(sprintf("KILL %d", ID), ...); mysql_close();
}
```
from bindings.cc 
and then `_handle.killConnection(id)` from js

rather than using the creating Client(js) from javascript side.

**Comment by mscdex** [comment]
@devzer01 The reason for creating a separate connection is because if the server is "stuck" (depending on what state it is in), it may not be able to process the kill request.

**Comment by devzer01** [comment]
@mscdex did you get a chance to review the PR.

**Comment by devzer01** [comment]
@mscdex Brian? can we merge this?

File: lib/Client.js [hunk 1]
```diff
@@ -567,8 +581,22 @@ Client.prototype._onclose = function(err) {
...
+  /**
+   there was a serious memory leak issue when using Client in a pool setup. 
+   if a client was used once and then close()ed the _handle was causing Client to never be garbage collection correctly. 
+   so the bindings was leaking around xMB everytime. 
+   the details around this is documented here 
+   https://github.com/mscdex/node-mariasql/pull/130 
+   and
+   https://github.com/mscdex/node-mariasql/pull/133
+  **/
+  process.nextTick(this._clearHandle);
```
Comment (mscdex) [inline lib/Client.js]: This won't work since the `this` context will be lost when `_clearHandle()` is called.

File: lib/Client.js [hunk 2]
```diff
@@ -567,8 +581,22 @@ Client.prototype._onclose = function(err) {
   if (!err)
     this.emit('end');
   this.emit('close');
+  /**
+   there was a serious memory leak issue when using Client in a pool setup. 
+   if a client was used once and then close()ed the _handle was causing Client to never be garbage collection correctly.
```
Comment (mscdex) [inline lib/Client.js]: This line needs be wrapped at 80 columns.

**Comment by mscdex** [comment]
A couple of issues but otherwise LGTM I think.

**Comment by mscdex** [comment]
I've landed this in e04f62a27ca11a87b8273608014d257e29f0897b with a couple of tweaks. Let me know if this still works for you.

**Pull request closed without merging**